bob-access/src/authenticator/basic.rs

```diff
@@ -0,0 +1,72 @@
+use std::{collections::HashMap, net::IpAddr};
+
+use crate::{credentials::Credentials, error::Error};
+
+use super::{users_storage::UsersStorage, Authenticator};
+
+#[derive(Debug, Default, Clone)]
+pub struct Basic<Storage: UsersStorage> {
+    users_storage: Storage,
+    nodes: HashMap<IpAddr, Credentials>,
+}
+
+impl<Storage: UsersStorage> Basic<Storage> {
+    pub fn new(users_storage: Storage) -> Self {
+        Self {
+            users_storage,
+            nodes: HashMap::new(),
+        }
+    }
+
+    pub fn set_nodes_credentials(
+        &mut self,
+        nodes: HashMap<IpAddr, Credentials>,
+    ) -> Result<(), Error> {
+        if nodes
+            .values()
+            .all(|cred| cred.ip().is_some() && cred.username().is_some())
```

Comment:
Can you add some example, what do you mean?)

Reply:
I mean we have `CredentialsKind::Basic` and `CredentialsKind::Token`. Instead of using `username()` for all credentials, wouldn't it be better to work with `CredentialsKind` everywhere we need either username or password? In this case we can pattern match that `cred` is `Basic`.
This way if we add more `CredentialsKind` variants in the future, all places which relies on particular credential types will be visible with compile errors

Reply:
@Boneyan